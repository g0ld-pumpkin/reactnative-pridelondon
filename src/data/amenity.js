// @flow
import * as decode from "../lib/decode";
import type { Decoder } from "../lib/decode";

type AmenityType = "Toilet" | "First Aid";

const amenityTypeDecoder: Decoder<AmenityType> = decode.oneOf([
  decode.value("Toilet"),
  decode.value("First Aid")
]);

export type Amenity = {
  // important to keep this at the top level so type refinement works
  contentType: "amenity",
  id: string,
  locale: string,
  revision: number,
  fields: {
    type: AmenityType,
    location: { lat: number, lon: number }
  }
};

const decodeAmenity = (locale: string): Decoder<Amenity> =>
  decode.shape({
    contentType: decode.at(
      ["sys", "contentType", "sys", "id"],
      decode.value("amenity")
    ),
    id: decode.at(["sys", "id"], decode.string),
    locale: decode.succeed(locale),
    revision: decode.at(["sys", "revision"], decode.number),
    fields: decode.field(
      "fields",
      decode.shape({
        type: decode.at(["type", locale], amenityTypeDecoder),
        location: decode.at(
          ["location", locale],
          decode.shape({
            lat: decode.field("lat", decode.number),
            lon: decode.field("lon", decode.number)
          })
        )
      })
    )
  });

export default decodeAmenity;
