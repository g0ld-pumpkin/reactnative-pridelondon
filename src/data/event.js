// @flow
import R from "ramda";
import {
  set as setDate,
  diff as diffDate,
  add as addToDate,
  toFormat as formatDate,
  FORMAT_EUROPEAN_DATE,
  FORMAT_CONTENTFUL_ISO
} from "../lib/date";
import type { Maybe } from "../lib/maybe";
import * as maybe from "../lib/maybe";
import type { Decoder } from "../lib/decode";
import * as decode from "../lib/decode";
import type { FieldRef } from "./field-ref";
import decodeFieldRef from "./field-ref";
import * as colors from "../constants/colors";

export type Events = {
  [id: string]: Event
};

export type EventCategoryName =
  | "Cabaret and Variety"
  | "Community"
  | "Talks and Debates"
  | "Film and Screenings"
  | "Plays and Theatre"
  | "Social and Networking"
  | "Nightlife"
  | "Exhibition and Tours"
  | "Sports and Activities"
  | "Health"
  | "Music";

export const eventCategoryNames = [
  "Cabaret and Variety",
  "Community",
  "Talks and Debates",
  "Film and Screenings",
  "Plays and Theatre",
  "Social and Networking",
  "Nightlife",
  "Exhibition and Tours",
  "Sports and Activities",
  "Health",
  "Music"
];

const decodeEventCategoryName = decode.oneOf(
  eventCategoryNames.map(decode.value)
);

export type EventCategory = {
  label: EventCategoryName,
  color: string,
  contrast: boolean
};

// We use a switch here, rather than a look up so that we can ensure that
// this function will always return an EventCategory.
export const getEventCategoryFromName = (
  category: EventCategoryName
): EventCategory => {
  switch (category) {
    case "Cabaret and Variety":
      return {
        label: "Cabaret and Variety",
        color: colors.coralColor,
        contrast: false
      };
    case "Community":
      return {
        label: "Community",
        color: colors.brightLightBlueColor,
        contrast: true
      };
    case "Talks and Debates":
      return {
        label: "Talks and Debates",
        color: colors.darkSkyBlueColor,
        contrast: false
      };
    case "Film and Screenings":
      return {
        label: "Film and Screenings",
        color: colors.lightTealColor,
        contrast: true
      };
    case "Plays and Theatre":
      return {
        label: "Plays and Theatre",
        color: colors.warmPinkColor,
        contrast: false
      };
    case "Social and Networking":
      return {
        label: "Social and Networking",
        color: colors.turquoiseBlueColor,
        contrast: true
      };
    case "Nightlife":
      return {
        label: "Nightlife",
        color: colors.yellowColor,
        contrast: true
      };
    case "Exhibition and Tours":
      return {
        label: "Exhibition and Tours",
        color: colors.brightPurpleColor,
        contrast: false
      };
    case "Sports and Activities":
      return {
        label: "Sports and Activities",
        color: colors.vomitYellowColor,
        contrast: true
      };
    case "Health":
      return {
        label: "Health",
        color: colors.bubblegumPinkColor,
        contrast: true
      };
    case "Music":
      return {
        label: "Music",
        color: colors.cornflowerBlueColor,
        contrast: false
      };
    default:
      // Default case with cast to enable flow exhaustive match. This will
      // cause flow to error if passed anything other than an EventCategory;
      return (category: empty);
  }
};

export type Event = {
  // important to keep this at the top level so type refinement works
  contentType: "event",
  id: string,
  locale: string,
  revision: number,
  fields: {
    name: string,
    eventCategories: Array<EventCategoryName>,
    audience: Array<string>,
    startTime: string,
    endTime: string,
    location: { lat: number, lon: number },
    addressLine1: Maybe<string>,
    addressLine2: Maybe<string>,
    city: Maybe<string>,
    postcode: Maybe<string>,
    locationName: string,
    eventPriceLow: number,
    eventPriceHigh: number,
    accessibilityOptions: Array<string>,
    eventDescription: string,
    accessibilityDetails: Maybe<string>,
    email: Maybe<string>,
    phone: Maybe<string>,
    ticketingUrl: Maybe<string>,
    venueDetails: Array<string>,
    individualEventPicture: FieldRef,
    eventsListPicture: FieldRef,
    performances: Array<FieldRef>,
    recurrenceDates: Array<string>,
    stage: boolean
  }
};

export type SavedEvents = Set<string>;

export type EventDays = Event[][];

const maybeField = <A>(
  locale: string,
  field: string,
  decoder: Decoder<A>
): Decoder<Maybe<A>> =>
  decode.field(field, decode.maybe(decode.field(locale, decoder)));

const maybeFieldWithDefault = <A>(
  locale: string,
  field: string,
  decoder: Decoder<A>,
  defaultValue: A
): Decoder<A> =>
  decode.map(
    maybe.withDefault(defaultValue),
    maybeField(locale, field, decoder)
  );

export const decodeEvent = (locale: string): Decoder<Event> =>
  decode.shape({
    contentType: decode.at(
      ["sys", "contentType", "sys", "id"],
      decode.value("event")
    ),
    id: decode.at(["sys", "id"], decode.string),
    locale: decode.succeed(locale),
    revision: decode.at(["sys", "revision"], decode.number),
    fields: decode.field(
      "fields",
      decode.shape({
        name: decode.at(["name", locale], decode.string),
        eventCategories: decode.at(
          ["eventCategories", locale],
          decode.array(decodeEventCategoryName)
        ),
        audience: maybeFieldWithDefault(
          locale,
          "audience",
          decode.array(decode.string),
          []
        ),
        // may want to combine startTime and endTime to fix the
        // date flipping issue in the decoder
        // may also want to convert it to a DateTime
        startTime: decode.at(["startTime", locale], decode.string),
        endTime: decode.at(["endTime", locale], decode.string),
        location: decode.at(
          ["location", locale],
          decode.shape({
            lat: decode.field("lat", decode.number),
            lon: decode.field("lon", decode.number)
          })
        ),
        addressLine1: maybeField(locale, "addressLine1", decode.string),
        addressLine2: maybeField(locale, "addressLine2", decode.string),
        city: maybeField(locale, "city", decode.string),
        postcode: maybeField(locale, "postcode", decode.string),
        locationName: decode.at(["locationName", locale], decode.string),
        eventPriceLow: decode.at(["eventPriceLow", locale], decode.number),
        eventPriceHigh: decode.at(["eventPriceHigh", locale], decode.number),
        accessibilityOptions: maybeFieldWithDefault(
          locale,
          "accessibilityOptions",
          decode.array(decode.string),
          []
        ),
        eventDescription: decode.at(
          ["eventDescription", locale],
          decode.string
        ),
        accessibilityDetails: maybeField(
          locale,
          "accessibilityDetails",
          decode.string
        ),
        email: maybeField(locale, "email", decode.string),
        phone: maybeField(locale, "phone", decode.string),
        ticketingUrl: maybeField(locale, "ticketingUrl", decode.string),
        venueDetails: maybeFieldWithDefault(
          locale,
          "venueDetails",
          decode.array(decode.string),
          []
        ),
        individualEventPicture: decode.at(
          ["individualEventPicture", locale],
          decodeFieldRef
        ),
        eventsListPicture: decode.at(
          ["eventsListPicture", locale],
          decodeFieldRef
        ),
        performances: maybeFieldWithDefault(
          locale,
          "performances",
          decode.array(decodeFieldRef),
          []
        ),
        recurrenceDates: maybeFieldWithDefault(
          locale,
          "recurrenceDates",
          decode.array(decode.string),
          []
        ),
        stage: maybeFieldWithDefault(locale, "stage", decode.boolean, false)
      })
    )
  });

const formatEuropeanDate = value => formatDate(value, FORMAT_EUROPEAN_DATE);

const generateRecurringEvent = (event: Event) => (
  recurrenceStartTime: string
) => {
  const { endTime, startTime } = event.fields;

  const difference = diffDate(recurrenceStartTime, startTime);
  const recurrenceEndTime = addToDate(endTime, difference);

  return R.mergeDeepRight(event, {
    fields: {
      startTime: formatDate(recurrenceStartTime, FORMAT_CONTENTFUL_ISO),
      endTime: formatDate(recurrenceEndTime, FORMAT_CONTENTFUL_ISO),
      recurrenceDates: [
        formatEuropeanDate(startTime),
        ...event.fields.recurrenceDates
      ]
    },
    id: `${event.id}-recurrence-${formatEuropeanDate(recurrenceStartTime)}`
  });
};

const recurrenceDateToStartTime = (originalStartTime: string) => (
  recurrence: string
) => {
  const [recurrenceDay, recurrenceMonth, recurrenceYear] = recurrence.split(
    "/"
  );

  return formatDate(
    setDate(originalStartTime, {
      year: recurrenceYear,
      month: recurrenceMonth,
      day: recurrenceDay
    }),
    FORMAT_CONTENTFUL_ISO
  );
};

export const expandRecurringEvents = (event: Event): Array<Event> => {
  const recurrenceStartTimes = [
    event.fields.startTime,
    ...event.fields.recurrenceDates.map(
      recurrenceDateToStartTime(event.fields.startTime)
    )
  ];

  const events = R.uniq(recurrenceStartTimes)
    .slice(1)
    .map(generateRecurringEvent(event));
  return [event, ...events];
};
