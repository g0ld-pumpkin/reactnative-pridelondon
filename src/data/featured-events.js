// @flow
import * as decode from "../lib/decode";
import type { Decoder } from "../lib/decode";
import type { FieldRef } from "./field-ref";
import decodeFieldRef from "./field-ref";

export type FeaturedEvents = {
  // important to keep this at the top level so type refinement works
  contentType: "featuredEvents",
  id: string,
  locale: string,
  revision: number,
  fields: {
    title: string,
    events: Array<FieldRef>
  }
};

const decodeFeaturedEvents = (locale: string): Decoder<FeaturedEvents> =>
  decode.shape({
    contentType: decode.at(
      ["sys", "contentType", "sys", "id"],
      decode.value("featuredEvents")
    ),
    id: decode.at(["sys", "id"], decode.string),
    locale: decode.succeed(locale),
    revision: decode.at(["sys", "revision"], decode.number),
    fields: decode.field(
      "fields",
      decode.shape({
        title: decode.at(["title", locale], decode.string),
        events: decode.at(["events", locale], decode.array(decodeFieldRef))
      })
    )
  });

export default decodeFeaturedEvents;
