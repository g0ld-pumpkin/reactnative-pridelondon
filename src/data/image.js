// @flow
import * as decode from "../lib/decode";
import type { Decoder } from "../lib/decode";

export type Images = {
  [id: string]: ImageDetails
};

export type ImageDetails = {
  id: string,
  revision: number,
  uri: string,
  width: number,
  height: number
};

export const getImageDetails = (images: Images) => (
  id: string,
  dimensions: ?{ width: number, height: number }
): ?ImageDetails => {
  const imageDetails = images[id];

  // Image details might be undefined in case of unpublished image
  if (!dimensions || !imageDetails) {
    return imageDetails;
  }

  const imageUriWithResizingBehaviourParameters = `${imageDetails.uri}?w=${
    dimensions.width
  }&h=${dimensions.height}&fit=fill`;
  return Object.assign({}, imageDetails, {
    uri: imageUriWithResizingBehaviourParameters
  });
};

export const decodeImageDetails = (locale: string): Decoder<ImageDetails> =>
  decode.shape({
    id: decode.at(["sys", "id"], decode.string),
    revision: decode.at(["sys", "revision"], decode.number),
    uri: decode.map(
      value => `https:${value}`,
      decode.at(["fields", "file", locale, "url"], decode.string)
    ),
    height: decode.at(
      ["fields", "file", locale, "details", "image", "height"],
      decode.number
    ),
    width: decode.at(
      ["fields", "file", locale, "details", "image", "width"],
      decode.number
    )
  });
