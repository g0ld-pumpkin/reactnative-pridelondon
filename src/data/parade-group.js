// @flow
import * as decode from "../lib/decode";
import type { Decoder } from "../lib/decode";
import type { Maybe } from "../lib/maybe";

export type ParadeGroup = {
  // important to keep this at the top level so type refinement works
  contentType: "paradeGroup",
  id: string,
  locale: string,
  revision: number,
  fields: {
    name: string,
    facebookUrl: Maybe<string>,
    twitterUrl: Maybe<string>,
    websiteUrl: Maybe<string>
  }
};

const maybeField = <A>(
  locale: string,
  field: string,
  decoder: Decoder<A>
): Decoder<Maybe<A>> =>
  decode.field(field, decode.maybe(decode.field(locale, decoder)));

const decodeParadeGroup = (locale: string): Decoder<ParadeGroup> =>
  decode.shape({
    contentType: decode.at(
      ["sys", "contentType", "sys", "id"],
      decode.value("paradeGroup")
    ),
    id: decode.at(["sys", "id"], decode.string),
    locale: decode.succeed(locale),
    revision: decode.at(["sys", "revision"], decode.number),
    fields: decode.field(
      "fields",
      decode.shape({
        name: decode.at(["name", locale], decode.string),
        facebookUrl: maybeField(locale, "facebookUrl", decode.string),
        twitterUrl: maybeField(locale, "twitterUrl", decode.string),
        websiteUrl: maybeField(locale, "websiteUrl", decode.string)
      })
    )
  });

export default decodeParadeGroup;
