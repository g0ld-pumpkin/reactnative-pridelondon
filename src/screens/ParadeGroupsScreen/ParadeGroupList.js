// @flow
import React, { Component } from "react";
import { StyleSheet, SectionList } from "react-native";
import type { SectionBase } from "react-native/Libraries/Lists/SectionList";
import type { ParadeGroup } from "../../data/parade-group";
import ParadeGroupDetails from "./ParadeGroupDetails";
import ContentPadding from "../../components/ContentPadding";
import SectionDivider from "../../components/SectionDivider";
import SectionHeader from "../../components/SectionHeader";
import Text from "../../components/Text";
import { whiteColor } from "../../constants/colors";
import text from "../../constants/text";

type Props = {
  paradeGroups: ParadeGroup[][],
  testID?: string
};

type Section = SectionBase<ParadeGroup> & { index: number };

type RenderItemInfo = {
  item: ParadeGroup, // eslint-disable-line react/no-unused-prop-types
  index: number,
  section: Section
};

type RenderSectionInfo = {
  section: Section // eslint-disable-line react/no-unused-prop-types
};

const isOnlyAlpha = RegExp("[A-Z]");

export const sectionTitle = (paradeGroup: ParadeGroup): string => {
  const leadingCharacter = paradeGroup.fields.name.charAt(0).toUpperCase();
  if (isOnlyAlpha.test(leadingCharacter)) {
    return leadingCharacter;
  }
  return "#";
};

const sections = (sectionedParadeGroups: ParadeGroup[][]): Section[] =>
  sectionedParadeGroups.map((paradeGroups, index) => ({
    data: paradeGroups,
    key: sectionTitle(paradeGroups[0]),
    index
  }));

const getId = (item: { id: string }): string => item.id;

class ParadeGroupList extends Component<Props> {
  static defaultProps = {
    testID: undefined
  };

  shouldComponentUpdate(nextProps: Props) {
    const { paradeGroups } = this.props;
    const { paradeGroups: nextParadeGroups } = nextProps;

    return paradeGroups !== nextParadeGroups;
  }

  itemSeparator = () => (
    <ContentPadding>
      <SectionDivider />
    </ContentPadding>
  );

  keyExtractor = getId;

  renderItem = ({ item }: RenderItemInfo) => (
    <ContentPadding>
      <ParadeGroupDetails paradeGroup={item} />
    </ContentPadding>
  );

  renderSectionHeader = ({ section }: RenderSectionInfo) => (
    <SectionHeader title={sectionTitle(section.data[0])} />
  );

  renderHeader = () => (
    <ContentPadding style={styles.descriptionContainer}>
      <Text type="h1">{text.paradeGroups.title}</Text>
      <Text style={styles.h2} type="h2" color="lightNavyBlueColor">
        {text.paradeGroups.subTitle}
      </Text>
      <Text style={styles.paragraph}>{text.paradeGroups.paragraph1}</Text>
      <Text style={styles.paragraph}>{text.paradeGroups.paragraph2}</Text>
    </ContentPadding>
  );

  render() {
    const { paradeGroups, testID } = this.props;

    return (
      <SectionList
        stickySectionHeadersEnabled
        sections={sections(paradeGroups)}
        renderSectionHeader={this.renderSectionHeader}
        renderItem={this.renderItem}
        keyExtractor={this.keyExtractor}
        contentContainerStyle={styles.container}
        ItemSeparatorComponent={this.itemSeparator}
        ListHeaderComponent={this.renderHeader}
        testID={testID}
        windowSize={10}
      />
    );
  }
}

const styles = StyleSheet.create({
  container: {
    paddingTop: 0,
    backgroundColor: whiteColor
  },
  descriptionContainer: {
    marginTop: 24,
    marginBottom: 16
  },
  h2: {
    marginBottom: 4
  },
  paragraph: {
    marginBottom: 8
  }
});

export default ParadeGroupList;
