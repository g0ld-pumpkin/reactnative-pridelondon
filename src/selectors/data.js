// @flow
import type { Event, Events } from "../data/event";
import type { FeaturedEvents } from "../data/featured-events";
import type { FieldRef } from "../data/field-ref";
import type { Performance, Performances } from "../data/performance";
import type { Amenity } from "../data/amenity";
import type { State as DataState } from "../reducers/data";

type ObjectWithId<A> = {
  id: string
} & A;

const reduceToMapHelp = <A>(
  acc: { [id: string]: ObjectWithId<A> },
  item: ObjectWithId<A>
): { [id: string]: ObjectWithId<A> } => {
  acc[item.id] = item; // intentional mutation as this happens in a reduce
  return acc;
};

export const selectLoading = (data: DataState): boolean => data.loading;

export const selectRefreshing = (data: DataState): boolean => data.refreshing;

export const selectEvents = (data: DataState): Array<Event> => data.events;

export const selectEventsMap = (events: Array<Event>): Events =>
  events.reduce(reduceToMapHelp, {});

export const selectEventById = (events: Events, id: string): ?Event =>
  events[id];

export const selectPerformancesMap = (data: DataState): Performances =>
  data.performances;

export const selectPerformanceById = (
  performances: Performances,
  id: string
): ?Performance => performances[id];

export const selectFeaturedEvents = (data: DataState): FeaturedEvents[] =>
  data.featuredEvents;

export const selectFeaturedEventsByTitle = (
  featuredEventsList: FeaturedEvents[],
  title: string
): ?FeaturedEvents =>
  featuredEventsList.find(entry => entry.fields.title === title);

export const selectAmenities = (data: DataState): Array<Amenity> =>
  data.amenities;

const resolveEventsHelp = (eventMap: Events) => (
  acc: Event[],
  reference: FieldRef
) => {
  const event = eventMap[reference.sys.id];
  if (event) {
    acc.push(event);
  }
  return acc;
};

export const resolveEvents = (
  eventMap: Events,
  references: FieldRef[]
): Event[] => references.reduce(resolveEventsHelp(eventMap), []);
