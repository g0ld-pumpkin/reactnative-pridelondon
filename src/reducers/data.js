// @flow
import R from "ramda";
import type { DataAction } from "../actions/data";
import type { Event } from "../data/event";
import type { FeaturedEvents } from "../data/featured-events";
import type { HeaderBanner } from "../data/header-banner";
import type { Images } from "../data/image";
import type { ParadeGroup } from "../data/parade-group";
import type { Performances } from "../data/performance";
import type { Sponsor } from "../data/sponsor";
import type { Amenity } from "../data/amenity";
import { decodeEvent, expandRecurringEvents } from "../data/event";
import decodeFeaturedEvents from "../data/featured-events";
import decodeHeaderBanner from "../data/header-banner";
import { decodeImageDetails } from "../data/image";
import decodeParadeGroup from "../data/parade-group";
import decodePerformance from "../data/performance";
import decodeSponsor from "../data/sponsor";
import decodeAmenity from "../data/amenity";
import locale from "../data/locale";
import type { Decoder } from "../lib/decode";
import { filterMap as decodeFilterMap, map as decodeMap } from "../lib/decode";
import { withDefault as resultWithDefault } from "../lib/result";

export type State = {
  events: Event[],
  featuredEvents: FeaturedEvents[],
  headerBanners: HeaderBanner[],
  images: Images,
  paradeGroups: ParadeGroup[],
  performances: Performances,
  sponsors: Sponsor[],
  amenities: Amenity[],
  loading: boolean,
  refreshing: boolean,
  noDataReceived: boolean
};

const defaultState = {
  events: [],
  featuredEvents: [],
  headerBanners: [],
  images: {},
  paradeGroups: [],
  performances: {},
  sponsors: [],
  amenities: [],
  loading: true,
  refreshing: false,
  noDataReceived: false
};

type ObjectWithId<A> = {
  id: string
} & A;

const reduceToMapHelp = <A>(
  acc: { [id: string]: ObjectWithId<A> },
  item: ObjectWithId<A>
): { [id: string]: ObjectWithId<A> } => {
  acc[item.id] = item; // intentional mutation as this happens in a reduce
  return acc;
};

// moving locale here so we can deal with it in a single place
// this can be moved inside the reducer function if we later want
// to make this dynamic
const decodeEvents: Decoder<Array<Event>> = decodeMap(
  events => R.unnest(events.map(expandRecurringEvents)),
  decodeFilterMap(decodeEvent(locale))
);

const decodeFeaturedEventsCollection: Decoder<
  Array<FeaturedEvents>
> = decodeFilterMap(decodeFeaturedEvents(locale));

const decodeHeaderBanners: Decoder<Array<HeaderBanner>> = decodeFilterMap(
  decodeHeaderBanner(locale)
);

const decodeImages: Decoder<Images> = decodeMap(
  images => images.reduce(reduceToMapHelp, {}),
  decodeFilterMap(decodeImageDetails(locale))
);

const decodeParadeGroups: Decoder<Array<ParadeGroup>> = decodeFilterMap(
  decodeParadeGroup(locale)
);

const decodePerformances: Decoder<Performances> = decodeMap(
  performances => performances.reduce(reduceToMapHelp, {}),
  decodeFilterMap(decodePerformance(locale))
);

const decodeSponsors: Decoder<Array<Sponsor>> = decodeFilterMap(
  decodeSponsor(locale)
);

const decodeAmenities: Decoder<Array<Amenity>> = decodeFilterMap(
  decodeAmenity(locale)
);

const reducer = (state: State = defaultState, action: DataAction) => {
  switch (action.type) {
    case "REQUEST_CMS_DATA":
      return {
        ...state,
        loading: true,
        refreshing: false
      };
    case "REQUEST_UPDATE_CMS_DATA":
      return {
        ...state,
        loading: false,
        refreshing: true
      };
    case "RECEIVE_CMS_DATA":
      return {
        loading: false,
        refreshing: false,
        noDataReceived: false,
        events: resultWithDefault([], decodeEvents(action.data.entries)),
        featuredEvents: resultWithDefault(
          [],
          decodeFeaturedEventsCollection(action.data.entries)
        ),
        headerBanners: resultWithDefault(
          [],
          decodeHeaderBanners(action.data.entries)
        ),
        images: resultWithDefault({}, decodeImages(action.data.assets)),
        paradeGroups: resultWithDefault(
          [],
          decodeParadeGroups(action.data.entries)
        ),
        performances: resultWithDefault(
          {},
          decodePerformances(action.data.entries)
        ),
        sponsors: resultWithDefault([], decodeSponsors(action.data.entries)),
        amenities: resultWithDefault([], decodeAmenities(action.data.entries))
      };
    case "NO_DATA_RECEIVED":
      return {
        ...state,
        loading: false,
        refreshing: false,
        noDataReceived: true
      };
    default:
      return state;
  }
};

export default reducer;
