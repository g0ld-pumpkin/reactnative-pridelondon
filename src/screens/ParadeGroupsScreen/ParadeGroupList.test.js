// @flow
import React from "react";
import { shallow } from "enzyme";
import ParadeGroupList, { sectionTitle } from "./ParadeGroupList";
import { generateParadeGroup, sampleOne } from "../../data/__test-data";

describe("ParadeGroupList", () => {
  const paradeGroups = [
    [sampleOne(generateParadeGroup, { seed: 5432 })],
    [sampleOne(generateParadeGroup, { seed: 4534 })]
  ];
  const render = props =>
    shallow(<ParadeGroupList paradeGroups={paradeGroups} {...props} />);

  it("renders correctly", () => {
    const output = render();
    expect(output).toMatchSnapshot();
  });

  it("renders header correctly", () => {
    const renderHeader = render().prop("ListHeaderComponent");
    const output = renderHeader();

    expect(output).toMatchSnapshot();
  });

  it("renders section headers correctly", () => {
    const renderSectionHeader = render().prop("renderSectionHeader");
    const output = renderSectionHeader({ section: { data: paradeGroups[0] } });

    expect(output).toMatchSnapshot();
  });

  it("renders items correctly", () => {
    const renderItem = render().prop("renderItem");
    const output = renderItem({
      item: paradeGroups[0][0],
      index: 0,
      section: {
        index: 0
      }
    });

    expect(output).toMatchSnapshot();
  });

  it("renders item separators correctly", () => {
    const ItemSeparatorComponent = render().prop("ItemSeparatorComponent");
    const output = shallow(<ItemSeparatorComponent />);

    expect(output).toMatchSnapshot();
  });

  it("extracts keys correctly", () => {
    const keyExtractor = render().prop("keyExtractor");
    const key = keyExtractor(paradeGroups[0][0]);

    expect(key).toBe(paradeGroups[0][0].id);
  });
});

describe("sectionTitle", () => {
  it("returns the first letter of the name in uppercase", () => {
    const paradeGroup = sampleOne(generateParadeGroup, { seed: 5432 });
    paradeGroup.fields.name = "a parade group name";
    expect(sectionTitle(paradeGroup)).toEqual("A");
  });

  it("returns # if the first letter of the name is numeric", () => {
    const paradeGroup = sampleOne(generateParadeGroup, { seed: 3245 });
    paradeGroup.fields.name = "7 Dwarves";
    expect(sectionTitle(paradeGroup)).toEqual("#");
  });

  it("returns # if the first letter of the name is a symbol", () => {
    const paradeGroup = sampleOne(generateParadeGroup, { seed: 3245 });
    paradeGroup.fields.name = "$$$";
    expect(sectionTitle(paradeGroup)).toEqual("#");
  });
});
