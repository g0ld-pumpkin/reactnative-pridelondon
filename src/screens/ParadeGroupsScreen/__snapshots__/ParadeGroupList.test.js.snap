// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParadeGroupList renders correctly 1`] = `
<SectionList
  ItemSeparatorComponent={[Function]}
  ListHeaderComponent={[Function]}
  contentContainerStyle={
    Object {
      "backgroundColor": "#ffffff",
      "paddingTop": 0,
    }
  }
  data={Array []}
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  renderSectionHeader={[Function]}
  scrollEventThrottle={50}
  sections={
    Array [
      Object {
        "data": Array [
          Object {
            "contentType": "paradeGroup",
            "fields": Object {
              "facebookUrl": "https://red-badger.com/CTB4f2a0Khyl0b1T4lOFk11H",
              "name": "1x2yj",
              "twitterUrl": "https://red-badger.com/1r59u1Oj4T7D55BUw",
              "websiteUrl": null,
            },
            "id": "d18Cp545J4C996o5",
            "locale": "en-GB",
            "revision": 1,
          },
        ],
        "index": 0,
        "key": "#",
      },
      Object {
        "data": Array [
          Object {
            "contentType": "paradeGroup",
            "fields": Object {
              "facebookUrl": "https://red-badger.com/U",
              "name": "99b3j372i",
              "twitterUrl": null,
              "websiteUrl": "https://red-badger.com/R7yfk3P4z96cTKJtb7",
            },
            "id": "642gdEs5xmkdO5v541Oau11NM7T",
            "locale": "en-GB",
            "revision": 1,
          },
        ],
        "index": 1,
        "key": "#",
      },
    ]
  }
  stickySectionHeadersEnabled={true}
  updateCellsBatchingPeriod={50}
  windowSize={10}
/>
`;

exports[`ParadeGroupList renders header correctly 1`] = `
<ContentPadding
  padding={Object {}}
  style={
    Object {
      "marginBottom": 16,
      "marginTop": 24,
    }
  }
>
  <Text
    color="blackColor"
    markdown={false}
    markdownStyle={Object {}}
    type="h1"
  >
    London Parade
  </Text>
  <Text
    color="lightNavyBlueColor"
    markdown={false}
    markdownStyle={Object {}}
    style={
      Object {
        "marginBottom": 4,
      }
    }
    type="h2"
  >
    Saturday 7 July from 12pm
  </Text>
  <Text
    color="blackColor"
    markdown={false}
    markdownStyle={Object {}}
    style={
      Object {
        "marginBottom": 8,
      }
    }
    type="text"
  >
    The Pride in London Parade is a platform for every part of Londonâ€™s LGBT+ community. These groups, nearly 500 in total, represent our most diverse Pride in London yet. From LGBT+ parent groups and student services to sports clubs and national charities, we are proud to give a platform to our community representatives, supporters and allies.
  </Text>
  <Text
    color="blackColor"
    markdown={false}
    markdownStyle={Object {}}
    style={
      Object {
        "marginBottom": 8,
      }
    }
    type="text"
  >
    Find out who you can expect to see on the day from the list below.
  </Text>
</ContentPadding>
`;

exports[`ParadeGroupList renders item separators correctly 1`] = `
<ContentPadding
  padding={Object {}}
  style={Object {}}
>
  <SectionDivider />
</ContentPadding>
`;

exports[`ParadeGroupList renders items correctly 1`] = `
<ContentPadding
  padding={Object {}}
  style={Object {}}
>
  <ParadeGroupDetails
    paradeGroup={
      Object {
        "contentType": "paradeGroup",
        "fields": Object {
          "facebookUrl": "https://red-badger.com/CTB4f2a0Khyl0b1T4lOFk11H",
          "name": "1x2yj",
          "twitterUrl": "https://red-badger.com/1r59u1Oj4T7D55BUw",
          "websiteUrl": null,
        },
        "id": "d18Cp545J4C996o5",
        "locale": "en-GB",
        "revision": 1,
      }
    }
  />
</ContentPadding>
`;

exports[`ParadeGroupList renders section headers correctly 1`] = `
<SectionHeader
  hasShadow={true}
  title="#"
/>
`;
